torch/csrc/jit/passes/concat_opt.cpp

```diff
@@ -0,0 +1,465 @@
+
+  void run() {
+    for (auto node : graph_->block()->nodes()) {
+      if (node->kind() == aten::cat) {
+        expandCat(node);
+      }
+    }
+    cleanupExpandedCatOps();
+    GRAPH_DUMP("Before reusing copy buffers: ", graph_);
+    reuseBuffersInCopies();
+  }
+
+ private:
+  // Expand cat node into multiple copy nodes.
+  //
+  // Example:
+  //     %2 = aten::clamp(%0, ...)
+  //     %3 = aten::clamp(%1, ...)
+  //     %10 = prim::ListConstruct(%2, %3)
+  //     %11 = aten::cat(%10, ...)
+  //     ...
+  //         = %11 ... // Use %11
+  //
+  // After expanding cat:
+  //     %2 = aten::clamp(%0, ...)
+  //     %3 = aten::clamp(%1, ...)
+  //     %20 = aten::empty(...)          // cat output buffer
+  //     %21 = aten::slice(%20, ...)     // slice for %2
+  //     %22 = aten::copy_(%21, %2)      // copy %2
+  //     %23 = aten::slice(%20, ...)     // slice for %3
+  //     %24 = aten::copy_(%23, %3)      // copy %3
+  //     ...
+  //         = %20 ... // Use %20 in place of %11
+  void expandCat(Node* node) {
+    GRAPH_DEBUG("Considering cat node for expansion: ", node);
+    // Do not optimize cat nodes whose inputs are mutated in the graph.
+    // TODO: Improve this by checking if it is mutated in the graph region
+    // where this optimization is applied.
+    if (isListMutatedInGraph(node->input(0), *graph_)) {
+      return;
+    }
+    if (!allShapesAreKnown(node)) {
+      // Can't expand when shapes are not known for the `cat` op.
+      return;
+    }
+    for (auto cat_inp : node->input(0)->node()->inputs()) {
+      if (!shapeIsKnown(cat_inp)) {
+        // Can't expand when shapes of the inputs to `cat` are not known.
+        return;
+      }
+    }
+    auto cat_dim = node->input(1);
+    if (!cat_dim->node()->hasAttribute(attr::value)) {
+      // Can't expand when cat dimension is not a constant.
+      return;
+    }
+    int cat_dim_value = cat_dim->node()->i(attr::value);
+
+    // Set the insertion point to the curent `cat` node.
+    graph_->setInsertPoint(node);
+    auto none = graph_->insertConstant(IValue());
+    auto one = graph_->insertConstant(1);
+    auto empty_symbol = Symbol::aten("empty");
+    auto slice_symbol = Symbol::aten("slice");
+    auto copy_symbol = Symbol::aten("copy_");
+
+    // Insert the constants needed for the `cat` output buffer size.
+    auto tensortype = dynamic_cast<TensorType*>(node->output()->type().get());
+    TORCH_INTERNAL_ASSERT(tensortype);
+    auto tensortype_sizes = tensortype->sizes();
+    std::vector<Value*> cat_out_size;
+    for (size_t i = 0; i < tensortype_sizes.size(); ++i) {
+      cat_out_size.push_back(graph_->insertConstant(tensortype_sizes[i]));
+    }
+
+    // Create a list of int for `cat` output buffer size.
+    auto cat_out_size_list = graph_->createList(IntType::get(), cat_out_size);
+    cat_out_size_list->insertBefore(node);
+
+    // Create an empty buffer to be used as `cat` output buffer.
+    auto cat_out_empty = graph_->create(
```

Comment:
So, this is inserting the `aten::empty(int[] size, *, int? dtype, int? layout, Device? device, bool? pin_memory, MemoryFormat? memory_format=contiguous_format) -> Tensor`. schema and not specifying any value other than the size.

This only works for cpu float tensors as written because that is the default tensor instantiated with `empty`, but would break on any non-float non-cpu inputs.

Reply:
What is the correct way to instantiate an `empty` tensor here?

Reply:
Ideally we'd have an operator that took in a Tensor and a size, and initialized a Tensor to that size. Until that exists you just have to manually insert nodes to get the `dtype/layout/device/memory_format` from one of the inputs. [Here](https://github.com/pytorch/pytorch/pull/52175/files#diff-c098125f9cb12f6523dba61c6eebd88ba0f49914261bc3879b83526c81b49991R45) is an example of doing that. 

This also begs another question: how does this optimization work with different memory_formats? What's the eager behavior on concat with different memory formats / non-contiguity ?

Reply:
I'm inclining towards restricting this optimization to only contiguous Tensors.